python-checks/src/main/java/org/sonar/python/checks/InstanceMethodSelfAsFirstCheck.java

```diff
@@ -0,0 +1,197 @@
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.FunctionDef;
+import org.sonar.plugins.python.api.tree.HasSymbol;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.Parameter;
+import org.sonar.plugins.python.api.tree.ParameterList;
+import org.sonar.plugins.python.api.tree.RegularArgument;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key="S5720")
+public class InstanceMethodSelfAsFirstCheck extends PythonSubscriptionCheck {
+
+  // We allow "_" as sometimes it is conventionally used to signal that self won't be used.
+  private static final List<String> ALLOWED_NAMES = Arrays.asList("self", "_");
+  private static final List<String> EXCEPTIONS = Arrays.asList("__init_subclass__", "__class_getitem__", "__new__");
+
+  private static final String DEFAULT_IGNORED_DECORATORS = "classproperty";
+  private List<String> decoratorsToExclude;
+
+  @RuleProperty(
+    key = "ignoredDecorators",
+    description = "Comma-separated list of decorators which will disable this rule.",
+    defaultValue = DEFAULT_IGNORED_DECORATORS)
+  private String ignoredDecorators = DEFAULT_IGNORED_DECORATORS;
+
+  private List<String> getExcludedDecorators() {
+    if (decoratorsToExclude == null) {
+      decoratorsToExclude = new ArrayList<>();
+      decoratorsToExclude.add("staticmethod");
+      decoratorsToExclude.add("classmethod");
+      decoratorsToExclude.addAll(Arrays.asList(this.ignoredDecorators.split(",")));
+    }
+
+    return decoratorsToExclude;
+  }
+
+  private boolean isNonInstanceMethodDecorator(Decorator decorator) {
+    String fqn = decorator.name().names().stream().map(Name::name).collect(Collectors.joining("."));
+    return this.getExcludedDecorators().stream().anyMatch(fqn::contains);
+  }
+
+  private static class OldStyleClassMethodVisitor extends BaseTreeVisitor {
+    private Symbol symbol;
+    private boolean isOldStyleClassMethod = false;
+
+    public OldStyleClassMethodVisitor(Symbol symbol) {
+      this.symbol = symbol;
+    }
+
+    @Override
+    public void visitCallExpression(CallExpression pyCallExpressionTree) {
+      Symbol calleeSymbol = pyCallExpressionTree.calleeSymbol();
+      if (calleeSymbol == null || !"classmethod".equals(calleeSymbol.name())) {
+        return;
+      }
+
+      this.isOldStyleClassMethod |= pyCallExpressionTree.arguments().stream().anyMatch(argument -> {
+        if (argument instanceof RegularArgument) {
+          Expression expression = ((RegularArgument) argument).expression();
+          if (expression instanceof HasSymbol) {
+            return Objects.equals(((HasSymbol) expression).symbol(), this.symbol);
+          }
+        }
+
+        return false;
+      });
+    }
+  }
+
+  private static boolean isUsedAsOldStyleClassMethod(ClassDef parent, Symbol symbol) {
+    OldStyleClassMethodVisitor visitor = new OldStyleClassMethodVisitor(symbol);
```

Comment:
One alternative way of knowing this would be to go through `Symbol.usages`

```java
symbol.usages().stream()
  .anyMatch(usage -> isDeclaredAsClassMethod(usage.tree))

private static boolean isDeclaredAsClassMethod(Tree tree) {
 // check if parent is call expression called "classmethod"
}
```
However, I'm fine with visitor implementation as well, I let you choose what you prefer :)